presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1536,41 +1523,73 @@ public void flush()
         public void close()
                 throws IOException
         {
+            if (failed) {
+                try {
+                    abortUpload();
+                    return;
+                }
+                catch (RuntimeException e) {
+                    throw new IOException(e);
+                }
+            }
+
             try {
                 flushBuffer(true);
                 waitForPreviousUploadFinish();
             }
             catch (IOException | RuntimeException e) {
-                abortUpload(true);
+                abortUploadSuppressed(e);
                 throw e;
             }
 
-            if (!parts.isEmpty()) {
-                try {
-                    finishUpload(parts);
-                }
-                catch (RuntimeException e) {
-                    abortUpload(true);
-                    throw new IOException(e);
-                }
+            try {
+                uploadId.ifPresent(this::finishUpload);
             }
-            else {
-                abortUpload(false);
+            catch (RuntimeException e) {
+                abortUploadSuppressed(e);
+                throw new IOException(e);
             }
         }
 
         private void flushBuffer(boolean finished)
                 throws IOException
         {
+            try {
+                waitForPreviousUploadFinish();
```

Comment:
It seems that we should wait inside the below if branches? Because we only wait to flush when the buffer is full or it's the last part.